plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange.hs

```diff
@@ -153,24 +148,41 @@ findPosition pos root = go Nothing root
             startOfRight <- _start . _codeRange_range <$> V.headM right
             if pos < startOfRight then binarySearchPos left else binarySearchPos right
 
-appendFoldingRanges  :: Maybe FoldingRange -> [Maybe FoldingRange] -> [Maybe FoldingRange]
-appendFoldingRanges a [] = [a]
-appendFoldingRanges a (x:xs) = x : appendFoldingRanges a xs
-
 findFoldingRanges :: CodeRange -> [Maybe FoldingRange]
-findFoldingRanges root = go [] root
-    where
-        -- Helper function for recursion. The range list is built top-down
-        go :: [Maybe FoldingRange] -> CodeRange -> [Maybe FoldingRange]
-        go acc node = do
-                let acc' = appendFoldingRanges (Just $ FoldingRange lineStart 0 lineEnd 0 0) acc
-                go (acc' children)
-            where
-                range = _codeRange_range node
-                children = _codeRange_children node
-                Range startPos endPos = range
-                Position lineStart _= startPos
-                Position lineEnd _ = endPos
+findFoldingRanges root = do
+    let acc = []
+    let node = _codeRange_children root
+    binarySearchFoldingRange node acc
+
+binarySearchFoldingRange :: Vector CodeRange -> [Maybe FoldingRange] -> [Maybe FoldingRange]
+binarySearchFoldingRange v acc
+    | V.null v = []
+    | V.length v == 1 = do
+        headOfCodeRange <- V.headM v
+        let foldingRangesOfRange = createFoldingRange headOfCodeRange
+        let acc' = acc ++ [foldingRangesOfRange]
+        acc'
+    | otherwise = do
+        let (left, right) = V.splitAt (V.length v `div` 2) v
+        _ <- binarySearchFoldingRange left acc
+        binarySearchFoldingRange right acc
+
+createFoldingRange :: CodeRange -> Maybe FoldingRange
+createFoldingRange node1 = do
+    let range = _codeRange_range node1
+    -- let children = _codeRange_children node1
+    let Range startPos endPos = range
+    let Position lineStart _= startPos
+    let Position lineEnd _ = endPos
+    let codeRangeKind = _codeRange_kind node1
+
+    if codeRangeKind == CodeKindComment
```

Comment:
Tip: You can use pattern matching instead of `==`

Reply:
```
case codeRangeKind of
   CodeKindComment -> ...
   CodeKindImports -> ...
   ...
```

Reply:
I'd also be tempted to pull out the construction of the range, and just have this return the kind, i.e. 
```
let frk <- case codeRangeKind of
    CodeKindComment -> Just FoldingRangeComment
    ...
    _ -> Nothing
case frk of
    Just _ -> Just (FoldingRange lineStart Nothing lineEnd Nothing frk)
    Nothing -> Nothing
```

Reply:
I made a separate function that converts crk to frk, the function looks a lot cleaner now.